Framework/Muon/src/MuonGroupingAsymmetry.cpp

```diff
@@ -0,0 +1,318 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidMuon/MuonGroupingAsymmetry.h"
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidAPI/WorkspaceGroup.h"
+#include "MantidDataObjects/TableWorkspace.h"
+#include "MantidHistogramData/HistogramMath.h"
+#include "MantidKernel/ArrayProperty.h"
+#include "MantidKernel/make_unique.h"
+#include "MantidMuon/MuonAlgorithmHelper.h"
+
+#include <algorithm>
+
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Kernel;
+
+namespace {
+
+bool is_alnum_underscore(char c) {
+  return (isalpha(c) || isdigit(c) || (c == '_'));
+}
+
+bool checkPeriodInWorkspaceGroup(const int &period,
+                                 WorkspaceGroup_sptr workspace) {
+  return period <= workspace->getNumberOfEntries();
+}
+
+/**
+ * Estimate the asymmetrey for the given workspace (TF data).
+ * @param inputWS :: [input] Workspace to calculate asymmetry for
+ * @param index :: [input] GroupIndex (fit only the requested spectrum): use -1
+ * for "unset"
+ * @returns Result of the removal
+ */
+MatrixWorkspace_sptr estimateAsymmetry(const Workspace_sptr &inputWS,
+                                       const int index, const double startX,
+                                       const double endX) {
+  IAlgorithm_sptr asym =
+      AlgorithmManager::Instance().create("EstimateMuonAsymmetryFromCounts");
+  asym->setChild(true);
+  asym->setProperty("InputWorkspace", inputWS);
+  asym->setProperty("WorkspaceName", inputWS->getName());
+  if (index > -1) {
+    std::vector<int> spec(1, index);
+    asym->setProperty("Spectra", spec);
+  }
+  asym->setProperty("OutputWorkspace", "__NotUsed__");
+  asym->setProperty("StartX", startX);
+  asym->setProperty("EndX", endX);
+  asym->setProperty("NormalizationIn", 0.0);
+  asym->setProperty("OutputUnNormData", false);
+  asym->setProperty("OutputUnNormWorkspace", "tmp_unNorm");
+  asym->execute();
+  MatrixWorkspace_sptr outWS = asym->getProperty("OutputWorkspace");
+
+  return outWS;
+}
+
+Mantid::API::MatrixWorkspace_sptr
+calculateMuonAsymmetry(WorkspaceGroup_sptr inputWS,
```

Comment:
this is not actually a calculation and is an estimate, so we should change the name of the method